src/main/java/org/sonarlint/intellij/finding/persistence/LiveFindingCache.java

```diff
@@ -119,26 +120,22 @@ public Collection<Trackable> getPreviousFindings(VirtualFile file) {
    * Read findings from a file that are cached. On cache miss, it won't fallback to the persistent store.
    */
   @CheckForNull
-  public synchronized Collection<T> getLive(VirtualFile virtualFile) {
+  public Collection<T> getLive(VirtualFile virtualFile) {
     var liveFindings = cache.get(virtualFile);
     if (liveFindings != null) {
-      // Create a copy to avoid concurrent modification issues
+      // Create a copy to avoid concurrent modification issues later
       return new ArrayList<>(liveFindings);
     }
     return null;
   }
 
-  public synchronized void insertFinding(VirtualFile virtualFile, T finding) {
-    cache.computeIfAbsent(virtualFile, f -> new ArrayList<>()).add(finding);
-  }
-
   /**
-   * Flushes all cached entries to disk.
+   * Flushes all provided entries to disk.
    * It does not clear the cache.
    */
-  private void flushAll() {
+  private void flushAll(Map<VirtualFile, Collection<T>> newFindingsPerFile) {
```

Comment:
Before, this method was never called concurrently, because of the `synchronized` on the `replaceFindings` method. Now it can be called concurrently. Is it safe? Is `persistence.save(key, liveFindings);` thread safe?

Reply:
I pass the map as a parameter to avoid iterating over the cache, which could cause concurrency issues.
About the persistence, in fact we already have synchronization on `persistence.save(key, liveFindings)` but it might be a problem as it contains a read action. I will rework that